domain/status/state/state.go

```diff
@@ -242,6 +242,126 @@ ON CONFLICT(application_uuid) DO UPDATE SET
 	return nil
 }
 
+// GetRelationStatus gets the status of the given relation. It returns an error
+// satisfying [statuserrors.RelationNotFound] if the relation doesn't exist.
+func (st *State) getRelationStatus(
+	ctx context.Context,
+	tx *sqlair.TX,
+	uuid corerelation.UUID,
+) (status.StatusInfo[status.RelationStatusType], error) {
+	empty := status.StatusInfo[status.RelationStatusType]{}
+	id := relationUUID{
+		RelationUUID: uuid,
+	}
+	var sts relationStatus
+
+	stmt, err := st.Prepare(`
+SELECT &relationStatus.*
+FROM   relation_status
+WHERE  relation_uuid = $relationUUID.relation_uuid
+`, id, sts)
+	if err != nil {
+		return empty, errors.Capture(err)
+	}
+
+	err = tx.Query(ctx, stmt, id).Get(&sts)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return empty, statuserrors.RelationNotFound
+	} else if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return empty, errors.Capture(err)
+	}
+
+	statusType, err := status.DecodeRelationStatus(sts.StatusID)
+	if err != nil {
+		return empty, errors.Capture(err)
+	}
+	return status.StatusInfo[status.RelationStatusType]{
+		Status:  statusType,
+		Message: sts.Reason,
+		Since:   sts.Since,
+	}, nil
+}
+
+// SetRelationStatus saves the given relation status, overwriting any current
+// status data. It can return the following errors:
+//   - [statuserrors.RelationNotFound] if the relation doesn't exist.
+//   - [statuserrors.RelationStatusTransitionNotValid] if the current relation
+//     status cannot transition to the new relation status. the relation does
+//     not exist.
+func (st *State) SetRelationStatus(
+	ctx context.Context,
+	relationUUID corerelation.UUID,
+	sts status.StatusInfo[status.RelationStatusType],
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		// Get current status.
+		currentStatus, err := st.getRelationStatus(ctx, tx, relationUUID)
+		if err != nil {
+			return errors.Errorf("getting current relation status: %w", err)
+		}
+
+		// Check we can transition from current status to the new status.
+		err = status.RelationStatusTransitionValid(currentStatus, sts)
+		if err != nil {
+			return errors.Capture(err)
+		}
+
+		// If transitioning from Suspending to Suspended, retain any existing
+		// message so that any previous reason for suspending is retained.
```

Comment:
```suggestion
		// If transitioning from Suspending to Suspended, retain any existing
		// message so that any previous reason for suspending is retained if
		// the new message is empty.
```